### PR TITLE
UPF to Tarski Converter

### PR DESCRIPTION
**Pull request by Framba-Luca**

### CODE REVIEW

File: upf/solvers/upf_tarski_converter.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+        for ut in problem.user_types(): #adding user_types to the language
+            lang.sort(ut.name()) # type: ignore
+        for fluent in problem.fluents(): #adding fluents to the language
+            signature = []
+            for type in fluent.signature():
+                if type.is_user_type():
+                    signature.append(lang.get_sort(type.name())) # type: ignore
+                else:
+                    #typename will be the name that this type has in the tarski language
+                    typename = str(type).replace(' ','')
```
Comment (alvalentini) [inline upf/solvers/upf_tarski_converter.py:137]: We need to consider int and real types with no bounds

File: upf/solvers/upf_tarski_converter.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+                    if not lang.has_sort(typename):
+                        # the type is not in the language, therefore it must be added
+                        if type.is_int_type() or type.is_real_type():
```
Comment (alvalentini) [inline upf/solvers/upf_tarski_converter.py:140]: Keep them separate

File: upf/solvers/upf_tarski_converter.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+                            lang.interval(typename, lang.Real, type.lower_bound(), type.upper_bound()) # type: ignore
+                        else:
+                            raise NotImplementedError
+                    signature.append(lang.get_sort(typename))
+            if fluent.type().is_bool_type():
+                lang.predicate(fluent.name(), *signature)
+            else:
+                typename = str(fluent.type()).replace(' ','')
```
Comment (alvalentini) [inline upf/solvers/upf_tarski_converter.py:148]: Same as above. Use `lang.Real()` and 'land.Integer()' when there are no bounds.

File: upf/model/action.py [hunk 4]
```diff
@@ -139,6 +139,8 @@ def __repr__(self) -> str:
     def __eq__(self, oth: object) -> bool:
         if isinstance(oth, InstantaneousAction):
             cond = self._env == oth._env and self._name == oth._name and self._parameters == oth._parameters
+            if not (cond and set(self._preconditions) == set(oth._preconditions) and set(self._effects) == set(oth._effects)):
```
Comment (alvalentini) [inline upf/model/action.py:142]: Remove all debug prints

File: upf/interop/tarski.py [hunk 5]
```diff
@@ -133,7 +158,19 @@ def convert_tarski_problem(env: Environment, tarski_problem: tarski.fstrips.Prob
         signature = []
         for t in p.domain:
             signature.append(types[str(t.name)])
-        fluent = upf.model.Fluent(p.name, tm.RealType(), signature)
+        func_sort = p.sort[-1]
+        fluent = None # type: ignore
+        if isinstance(func_sort, Interval):
```
Comment (alvalentini) [inline upf/interop/tarski.py:163]: As discussed, here we can treat `func_sort == 'Integer'` and `func_sort == 'Real'` as special cases

File: upf/solvers/upf_tarski_converter.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+    def walk_bool_constant(self, expression: 'upf.model.FNode', args: List['tarski.syntax.formulas.Formula']) -> 'tarski.syntax.formulas.Formula':
+        assert len(args) == 0
+        if expression.bool_constant_value():
+            return tarski.syntax.top
+        return tarski.syntax.bot
+
+    def walk_int_constant(self, expression: 'upf.model.FNode', args: List['tarski.syntax.formulas.Formula']) -> 'tarski.syntax.formulas.Formula':
+        assert len(args) == 0
+        # NOTE expression.int_constant_value() might need to be changed to str(expression.int_constant_value())
+        return tarski.syntax.Constant(expression.int_constant_value(), self.lang.Real) # type: ignore
```
Comment (alvalentini) [inline upf/solvers/upf_tarski_converter.py:102]: Why `lang.Real`?

File: upf/interop/from_tarski.py [hunk 7]
```diff
@@ -181,6 +242,8 @@ def convert_tarski_problem(env: Environment, tarski_problem: tarski.fstrips.Prob
             default_value = em.Real(Fraction(0))
         elif fluent.type().is_int_type():
             default_value = em.Int(0)
+        elif fluent.type().is_user_type():
+            continue #NOTE: not sure
```
Comment (alvalentini) [inline upf/interop/from_tarski.py:246]: Check this note and fix it

File: upf/interop/to_tarski.py [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+                            parameters,
+                            precondition=tfc.convert_formula(em.And(action.preconditions())),
+                            effects=[_convert_effect(e, tfc, em) for e in action.effects()])
+    for fluent_exp, value_exp in problem.initial_values().items():
+        if value_exp.is_bool_constant():
+            if value_exp.constant_value():
+                parameters = []
+                for a in fluent_exp.args():
+                    if not a.is_object_exp():
+                        raise #NOTE is this raise right? If signature has ints/reals not
```
Comment (alvalentini) [inline upf/interop/to_tarski.py:181]: Fix this

File: upf/interop/to_tarski.py [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+            assert type.upper_bound() is None # type: ignore #must be true, otherwise branch would be skipped
+            typename = 'Natural'
+        else:
+            raise upf.exceptions.UPFProblemDefinitionError('Int type with just one bound is not accepted by tarski')
+    elif type.is_real_type() and (type.lower_bound() is None or type.upper_bound() is None): # type: ignore
+        if type.lower_bound() is None and type.upper_bound() is None: # type: ignore
+            typename = 'Real'
+        else:
+            raise upf.exceptions.UPFProblemDefinitionError('Real type with just one bound is not accepted by tarski')
+    #NOTE: Still missing the part where fluents can be of type user_type, not sure if this code works for it
```
Comment (alvalentini) [inline upf/interop/to_tarski.py:211]: Check this comment

File: upf/walkers/type_checker.py [hunk 10]
```diff
@@ -35,15 +35,20 @@ def get_type(self, expression: FNode) -> 'upf.model.types.Type':
...
                 (t_left.is_real_type() and t_right.is_real_type()) or
                 (t_left.is_real_type() and t_right.is_int_type())):
+
+            print('AAAAAAA')
+            print(t_left)
+            print(t_right)
+            print(t_left.is_int_type())
+            print(t_right.is_int_type())
+            print('BBBBBBB')
+
```
Comment (alvalentini) [inline upf/walkers/type_checker.py:51]:
```suggestion
```

File: upf/walkers/type_checker.py [hunk 11]
```diff
@@ -69,7 +80,13 @@ def walk_fluent_exp(self, expression: FNode, args: List['upf.model.types.Type'])
...
             return None
         for (arg, p_type) in zip(args, f.signature()):
-            if arg != p_type:
+            if not self.is_compatible_type(arg, p_type):
+
+                print('SIGNATURE->')
+                print(f.signature())
+                print(f.signature()[0].is_int_type())
+                print(f.signature()[0].is_user_type())
+
```
Comment (alvalentini) [inline upf/walkers/type_checker.py:89]:
```suggestion
```

File: upf/interop/to_tarski.py [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+        tarski_param_rep = self.lang.variable(expression.parameter().name(), self.lang.get_sort(_type_name_added_to_language_if_needed(self.lang, expression.parameter().type()))) # type: ignore
+        return tarski_param_rep #NOTE :not sure
+
+    def walk_variable_exp(self, expression: 'upf.model.FNode', args: List['tarski.syntax.formulas.Formula']) -> 'tarski.syntax.formulas.Formula':
+        tarski_variable_rep = self.lang.variable(expression.variable().name(), self.lang.get_sort(_type_name_added_to_language_if_needed(self.lang, expression.variable().type()))) # type: ignore
+        return tarski_variable_rep #NOTE :not sure
+
+    def walk_object_exp(self, expression: 'upf.model.FNode', args: List['tarski.syntax.formulas.Formula']) -> 'tarski.syntax.formulas.Formula':
+        tarski_object_rep = self.lang.get_constant(expression.object().name())
+        return tarski_object_rep #NOTE :not sure
```
Comment (alvalentini) [inline upf/interop/to_tarski.py:137]: Why the note "not sure"?

**Review by alvalentini — CHANGES_REQUESTED** [review]

File: upf/interop/to_tarski.py [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+        assert len(args) == 2
+        return tarski.syntax.Atom(self.lang.get_predicate(tarski.syntax.BuiltinPredicateSymbol.LE), args)
+
+    def walk_lt(self, expression: 'upf.model.FNode', args: List['tarski.syntax.formulas.Formula']) -> 'tarski.syntax.formulas.Formula':
+        assert len(args) == 2
+        return tarski.syntax.Atom(self.lang.get_predicate(tarski.syntax.BuiltinPredicateSymbol.LT), args)
+
+    def walk_fluent_exp(self, expression: 'upf.model.FNode', args: List['tarski.syntax.formulas.Formula']) -> 'tarski.syntax.formulas.Formula':
+        tarski_fluent_rep = self.lang.get(expression.fluent().name())
+        new_args = [x.constant_value() if x.is_int_constant() else args[i] for i, x in enumerate(expression.args())]
```
Comment (alvalentini) [inline upf/interop/to_tarski.py:78]: This line is useless

File: upf/interop/to_tarski.py [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+        new_args = []
+        for i, x in enumerate(expression.args()):
+            if x.is_int_constant():
+                type = (expression.fluent().signature())[i]
+                typename = _type_name_added_to_language_if_needed(self.lang, type)
+                constant = tarski.syntax.Constant(x.int_constant_value(), \
+                    self.lang.get_sort(typename))
+                new_args.append(constant)
+            elif x.is_real_constant():
+                pass
```
Comment (alvalentini) [inline upf/interop/to_tarski.py:88]: Here we can raise an exception

File: upf/interop/to_tarski.py [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+        #add action to the problem
+        new_problem.action(action.name,
+                            parameters,
+                            precondition=tfc.convert_formula(em.And(action.preconditions())),
+                            effects=[_convert_effect(e, tfc, em) for e in action.effects()])
+    for fluent_exp, value_exp in problem.initial_values().items():
+        if value_exp.is_bool_constant():
+            if value_exp.constant_value():
+                parameters = []
+                for a in fluent_exp.args():#NOTE is this raise right? If signature has ints/reals not
```
Comment (alvalentini) [inline upf/interop/to_tarski.py:178]:
```suggestion
                for a in fluent_exp.args():
```

File: upf/io/pddl_reader.py [hunk 16]
```diff
@@ -105,7 +105,7 @@ def problem(self):
     @property
     def parameters(self):
         return self._parameters
-
+
```
Comment (alvalentini) [inline upf/io/pddl_reader.py:108]:
```suggestion

```

File: upf/test/test_a_tarski_grounder.py [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+from upf.model.problem_kind import full_classical_kind, full_numeric_kind
+from upf.plan import SequentialPlan, ActionInstance
+
+
+class TestGrounder(TestCase):
+    def setUp(self):
+        TestCase.setUp(self)
+        self.problems = get_example_problems()
+
+    #TODO: add skipif
```
Comment (alvalentini) [inline upf/test/test_a_tarski_grounder.py:33]: Fix it

File: upf/test/test_a_tarski_grounder.py [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+                self.assertNotEqual(planner, None)
+                grounded_plan = planner.solve(grounded_problem)
+                plan = rewrite_back_plan_function(grounded_plan)
+                for ai in plan.actions():
+                    a = ai.action()
+                    self.assertEqual(a, problem.action(a.name))
+                with PlanValidator(problem_kind=problem.kind()) as pv:
+                    self.assertTrue(pv.validate(problem, plan))
+
+    #TODO: write right skipif
```
Comment (alvalentini) [inline upf/test/test_a_tarski_grounder.py:50]: Fix it

File: upf/test/test_a_tarski_grounder.py [hunk 19]
```diff
@@ -0,0 +1,86 @@
+# Copyright 2021 AIPlan4EU project
```
Comment (alvalentini) [inline upf/test/test_a_tarski_grounder.py:1]: Rename the file

File: upf/transformers/grounder.py [hunk 20]
```diff
@@ -34,7 +34,9 @@ class Grounder(Transformer):
     resulting problem will not have lifted actions anymore,
     but only grounded actions.
     '''
-    def __init__(self, problem: Problem, name: str = 'grnd'):
+    def __init__(self, problem: Problem, name: str = 'grnd', \
+            grounding_actions_map: Optional[Dict[Action, List[Tuple[FNode, ...]]]] = None, \
+            grounding_fluents_map: Optional[Dict[Fluent, Tuple[FNode, ...]]] = None):
```
Comment (alvalentini) [inline upf/transformers/grounder.py:39]: This is not used, right?

File: upf/transformers/grounder.py [hunk 21]
```diff
@@ -70,29 +74,47 @@ def get_rewritten_problem(self) -> Problem:
...
+                    self._old_to_new[old_action] = [new_action]
+                else:
+                    # test that the action is in the dict of the actions that need to be in the grounded problem
+                    test = self._grounding_actions_map.get(old_action, None)
+                    if test is not None:
+                        new_action = old_action.clone()
+                        self._new_problem.add_action(new_action)
+                        self._new_to_old[new_action] = old_action
+                        self._map_parameters[new_action] = []
+                        self._old_to_new[old_action] = [new_action]
```
Comment (alvalentini) [inline upf/transformers/grounder.py:91]:
```suggestion
                if self._grounding_actions_map is None or \
                   self._grounding_actions_map.get(old_action, None) is not None:
                    new_action = old_action.clone()
                    self._new_problem.add_action(new_action)
                    self._new_to_old[new_action] = old_action
                    self._map_parameters[new_action] = []
                    self._old_to_new[old_action] = [new_action]
```

File: upf/solvers/tarski_grounder.py [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+        supported_kind.set_fluents_type('NUMERIC_FLUENTS') # type: ignore
+        supported_kind.set_effects_kind('INCREASE_EFFECTS') # type: ignore
+        supported_kind.set_effects_kind('DECREASE_EFFECTS') # type: ignore
+        return problem_kind <= supported_kind
+
+    def ground(self, problem: 'upf.model.Problem') -> Tuple['upf.model.Problem', Callable[[upf.plan.Plan], upf.plan.Plan]]:
+        tarski_problem = upf.interop.convert_problem_to_tarski(problem)
+        actions = None
+        state_variables = None
+        #TODO: understand when LP works and when not and use the problem_kind instead of try-catch
```
Comment (alvalentini) [inline upf/solvers/tarski_grounder.py:71]: Before calling the LPGroundingStrategy we should add a check like this:
```
gringo = shutil.which("gringo")
if gringo is None:
    raise CommandNotFoundError("gringo")
```

File: upf/solvers/tarski_grounder.py [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+        try:
+            lpgs = LPGroundingStrategy(tarski_problem)
+            actions = lpgs.ground_actions()
```
Comment (alvalentini) [inline upf/solvers/tarski_grounder.py:74]: This command can throw a `ReachabilityLPUnsolvable` exception. We should handle it explicitly.

**Review by alvalentini — CHANGES_REQUESTED** [review]

**Review by mikand — CHANGES_REQUESTED** [review]
Here are some comments

File: upf/solvers/tarski_grounder.py [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+        return problem_kind <= supported_kind
+
+    def ground(self, problem: 'upf.model.Problem') -> Tuple['upf.model.Problem', Callable[[upf.plan.Plan], upf.plan.Plan]]:
+        tarski_problem = upf.interop.convert_problem_to_tarski(problem)
+        actions = None
+        gringo = shutil.which('gringo')
+        if gringo is None:
+            raise tarski.errors.CommandNotFoundError('gringo not installed, try use sudo apt install gringo')
+        if problem.kind().has_continuous_numbers(): # type: ignore
+            raise upf.exceptions.UPFUsageError('tarski grounder can not be used for problems with continuous numbers.')
```
Comment (alvalentini) [inline upf/solvers/tarski_grounder.py:75]: This is implicit from the `supports` method. Not needed here.

File: upf/interop/to_tarski.py [hunk 25]
```diff
@@ -0,0 +1,238 @@
...
+        new_args = []
+        for i, x in enumerate(expression.args()):
+            if x.is_int_constant():
+                type = (expression.fluent().signature())[i]
+                typename = _type_name_added_to_language_if_needed(self.lang, type)
+                constant = tarski.syntax.Constant(x.int_constant_value(), \
+                    self.lang.get_sort(typename))
+                new_args.append(constant)
+            elif x.is_real_constant():
+                raise upf.exception.UPFProblemDefinitionError('Fluents can not have reals into their signatures.')
```
Comment (alvalentini) [inline upf/interop/to_tarski.py:87]:
```suggestion
                raise upf.exceptions.UPFProblemDefinitionError('Fluents can not have reals into their signatures.')
```

File: upf/test/test_tarski_grounder.py [hunk 26]
```diff
@@ -0,0 +1,84 @@
...
+import upf
+import upf.solvers
+
+from upf.shortcuts import *
+from upf.test import TestCase, skipIfNoOneshotPlannerForProblemKind
+from upf.test.examples import get_example_problems
+from upf.model.problem_kind import basic_classical_kind, full_classical_kind
+
+
+class TestGrounder(TestCase):
```
Comment (alvalentini) [inline upf/test/test_tarski_grounder.py:25]:
```suggestion
class TestTarskiGrounder(TestCase):
```

File: upf/test/test_tarski_converter.py [hunk 27]
```diff
@@ -0,0 +1,108 @@
...
+from upf.interop import convert_problem_from_tarski
+from upf.test import TestCase
+from upf.test.examples import get_example_problems
+from upf.interop import convert_problem_to_tarski
+from upf.model.problem_kind import full_classical_kind, full_numeric_kind
+from upf.plan import SequentialPlan, ActionInstance
+from upf.solvers import SequentialPlanValidator
+
+
+class TestGrounder(TestCase):
```
Comment (alvalentini) [inline upf/test/test_tarski_converter.py:28]:
```suggestion
class TestTarskiConverter(TestCase):
```

File: upf/solvers/tarski_grounder.py [hunk 28]
```diff
@@ -0,0 +1,97 @@
...
+        supported_kind.set_effects_kind('CONDITIONAL_EFFECTS') # type: ignore
+        supported_kind.set_fluents_type('OBJECT_FLUENTS') # type: ignore
+        return problem_kind <= supported_kind
+
+    def ground(self, problem: 'upf.model.Problem') -> Tuple['upf.model.Problem', Callable[[upf.plan.Plan], upf.plan.Plan]]:
+        tarski_problem = upf.interop.convert_problem_to_tarski(problem)
+        actions = None
+        gringo = shutil.which('gringo')
+        if gringo is None:
+            raise tarski.errors.CommandNotFoundError('gringo not installed, try use sudo apt install gringo')
```
Comment (alvalentini) [inline upf/solvers/tarski_grounder.py:69]:
```suggestion
            raise tarski.errors.CommandNotFoundError('gringo')
```

**Review by alvalentini — CHANGES_REQUESTED** [review]

**Review by alvalentini — APPROVED** [review]

File: upf/transformers/grounder.py [hunk 29]
```diff
@@ -70,29 +85,39 @@ def get_rewritten_problem(self) -> Problem:
...
+                    ds = domain_size(self._new_problem, t)
+                    domain_sizes.append(ds)
+                    ground_size *= ds
+                items_list: List[List[FNode]] = []
+                for size, type in zip(domain_sizes, type_list):
+                    items_list.append([domain_item(self._new_problem, type, j) for j in range(size)])
+                grounded_params_list = product(*items_list)
+            else:
+                # The grounding_actions_map is not None, therefore it must be used to ground
+                grounded_params_list = _create_generator_from_list_of_tuple(self._grounding_actions_map[old_action])
```
Comment (mikand) [inline upf/transformers/grounder.py:118]: There is no need for this function:
```suggestion
                grounded_params_list = iter(self._grounding_actions_map[old_action])
```

File: upf/transformers/grounder.py [hunk 30]
```diff
@@ -204,3 +236,7 @@ def get_fresh_name(self, original_name: str, parameters_names: Iterable[str] = [
             new_name = f'{base_name}_{str(count)}'
             count += 1
         return new_name
+
+def _create_generator_from_list_of_tuple(tuple_list: List[Tuple[FNode, ...]]):
```
Comment (mikand) [inline upf/transformers/grounder.py:240]: This function can be removed

**Review by mikand — APPROVED** [review]
@Framba-Luca Excellent work!

**Pull request merged**